### PR TITLE
SCUMM: v7-8: Implement pixel accurate text rendering

### PR DESCRIPTION
**Pull request by AndywinXp**
This PR is an expansion of the now closed [https://github.com/scummvm/scummvm/pull/3276](https://github.com/scummvm/scummvm/pull/3276), and consists for 95% of some amazing work done by @athrxx. 

The changes introduced in the PR:
- Reorganize the font, subtitles, texts, blastText, etc. handling in a better way, separating SCUMM v7-8 from the previous versions as much as it can;
- Remove duplicated text rendering code in the SMUSH module, by making use of the aforementioned text rendering routines;
- Produce pixel accurate rendering of any text object in every SCUMM v7-8 game.

As noted on the old PR, these changes properly fix issues in which the text is not wrapped correctly like in here:

With these changes, now ScummVM renders these texts correctly:

image

### CODE REVIEW

**Comment by AndywinXp** [comment]
DIG demo is now fixed; pretty ironically the answer lied in the FT disasm: `_haveMsg` should be set to 1 instead of 0xFF for SCUMM 7. The fix does not interfere with DIG (full), since it uses a different routine for escape codes.

**Comment by AndywinXp** [comment]
@athrxx having tested Full Throttle and the demo of The Dig from start to finish, this PR is now ready for a review (if and when you've got time for it, of course).

**Comment by athrxx** [comment]
Nice. I'll play through DIG and see how that goes...

**Comment by orgads** [comment]
cc @BLooperZ

**Comment by athrxx** [comment]
I have played through the English version of DIG without any text related issues.
I will give English COMI some testing, too...

**Comment by AndywinXp** [comment]
> I have played through the English version of DIG without any text related issues. I will give English COMI some testing, too...

Thanks!

**Comment by athrxx** [comment]
Just for the record, I noticed something weird in COMI Korean, but it is NOT due to this PR. I checked the original interpreter and it is exactly the same: The item description string at the bottom center will sometimes move up a bit, depending on whether the string contains a space character. This is due to the fact that the space characters are "drawn" with the normal font and have a font height of 40. Unlike the Korean font which only has a height of 16. So, as soon as there is a space character the engine will treat the whole string as height 40 instead of height 16 and move it up a bit to stay clear of the screen bottom.

I only put this here as a reminder that we have seen this and it is just fine as it is supposed to be...

**Comment by dwatteau** [comment]
Hi, for what it's worth, I've compared this with my original 2004 French CD interpreters.

* Full Throttle, French
    * compared against the `SPU(tm) version 7.3.5 (Dec 17 2002 00:46:23)` build (by Aaron Giles) on Windows XP
* The Dig, French
    * compared against the `SPU(tm) version 7.5.0 (Dec 17 2002 00:46:13)` build (by Aaron Giles) on Windows XP
    * the very first sentences of the game sound weird, and the audio of some dialogue is cut too early, but that's the same behavior as the original French game
* The Curse of Monkey Island, French
    * compared against the `SPU(tm) version 8.1.0 (Oct 27 1997 23:20:53)` build on Windows 98

I've mostly played the intros, and some random savegames. Everything looks identical between this ScummVM branch and my French interpreters (bug for bug, in the case of The Dig).

**Comment by AndywinXp** [comment]
> Hi, for what it's worth, I've compared this with my original 2004 French CD interpreters.
> 
> * Full Throttle, French
>   
>   * compared against the `SPU(tm) version 7.3.5 (Dec 17 2002 00:46:23)` build (by Aaron Giles) on Windows XP
> * The Dig, French
>   
>   * compared against the `SPU(tm) version 7.5.0 (Dec 17 2002 00:46:13)` build (by Aaron Giles) on Windows XP
>   * the very first sentences of the game sound weird, and the audio of some dialogue is cut too early, but that's the same behavior as the original French game
> * The Curse of Monkey Island, French
>   
>   * compared against the `SPU(tm) version 8.1.0 (Oct 27 1997 23:20:53)` build on Windows 98
> 
> I've mostly played the intros, and some random savegames. Everything looks identical between this ScummVM branch and my French interpreters (bug for bug, in the case of The Dig, and the bugs are not related to text rendering).

Thanks 😄 glad that it matches those particular interpreters since those are the exact ones I used for researching what needed to be done to complete athrxx work.

**Comment by AndywinXp** [comment]
@athrxx I've noticed some kind of regression wrt the master codebase: if you run ScummVM with parameters "-b 400 ft" (which is FT eng for me, doesn't matter if A or B), and get to the point where Malcolm Corley sings the song after the bathroom break, you'll notice that the text is considerably faster than it should be. 

This is related to the _haveMsg variable (escape code 3) and wasn't an issue in the master. In the master we handle it setting _haveMsg to 0xFF for all scumm versions, while here that doesn't appear to work (and indeed the disasm checks for _haveMsg == 1, which is what we do for FT in this PR).

Any idea why this happens? I admit I was pretty tired last time I checked but I recently couldn't find time to get back to it...

**Comment by athrxx** [comment]
> @athrxx I've noticed some kind of regression wrt the master codebase: if you run ScummVM from this PR with parameters "-b 400 ft" (which is FT eng for me, doesn't matter if A or B), and get to the point where Malcolm Corley sings the song after the bathroom break, you'll notice that the text is considerably faster than it should be.
> 
> This is related to the _haveMsg variable (escape code 3) and wasn't an issue in the master. In the master we handle it by setting _haveMsg to 0xFF for all scumm versions, while here that doesn't appear to work (and indeed the disasm checks for _haveMsg == 1, which is what we do for FT in this PR).
> 
> Any idea why this happens? I admit I was pretty tired last time I checked but I recently couldn't find time to get back to it...

Is this issue now fixed by your two latest commits? Or do I still need to do anything? :-)

**Comment by AndywinXp** [comment]
> Is this issue now fixed by your two latest commits? Or do I still need to do anything? :-)

I gave it a try earlier today but I couldn't find anything other than ScummVM-interpreter inconsistencies (which is what I fixed in the latest commits), so if you are able to, I really could use your help with that :(

**Comment by dwatteau** [comment]
@AndywinXp I think I've found a small regression in this new branch (built from commit ab8f7cc2b151b070828aece776d79925c01f7653).

1. Start The Dig from this branch
2. Press Esc until you can control Low
3. Open the Inventory
4. Use the PenUltimate, and then start the Asteroid Lander game
5. Don't play it, just let it fall

At least in the French version, the "Carburant" text (= Fuel) on the right now glitches: first the initial "C" letter is missing, and then as the object falls, this line will glitch (see the attached video). Regular ScummVM 2.5.1 and the Aaron Giles interpreter don't have this particular bug.

https://user-images.githubusercontent.com/9024526/160252129-3bc5b63f-77a8-424f-a26c-815f1dd0a2bd.mp4

**Comment by AndywinXp** [comment]
Nice catch! Might this be an issue related to the rect size of the text box? @athrxx

**Comment by athrxx** [comment]
Okay, I cannot reproduce it with English, Chinese, Korean or Hebrew, but if I just replace "Fuel :" with "Carburant:" it looks exactly like in the video.

The problem is that the script seems to send an x coordinate of 230 and the length of the French string is 101, so the string would exceed the screen width of 320.

Maybe the old style text printing does not clip the text like the newer version. Maybe it just has to be moved left so that it fits. But that's just guessing, I'll have to look at the disasm...

**Comment by AndywinXp** [comment]
image
Here's what it is supposed to look like (Aaron Giles interpreter 2002, the one sold on Steam, which uses the old text style), it prints the line using the coordinates "as-is" (i.e. the y is not adjusted to fit the screen). 

And thinking about it, this is what FT also did in a particular instance which made ScummVM crash before I added this fix [cc58ca0](https://github.com/scummvm/scummvm/pull/3733/commits/cc58ca00303b845fe57035c076368e18d77dfb90).

And I should add, if the fix to DIG's issue also solves the crash above, that could be the occasion to remove my almost-hacky fix altogether :-)

**Comment by athrxx** [comment]
It might be okay to remove that code block, since it seems to be wrong to adapt the x coord and I can't remember why I added that line in the first place.

The reason for the missing clipping can be seen in my comment in  CharsetRendererV7::drawChar(). I have simply not implemented it :-)

But I did implement it in NutRenderer::drawChar(), so it is probably a matter of copy/paste...

**Comment by athrxx** [comment]
Okay, we can't really clip as well as we do in the NutRenderer without modifying the classic renderer (which we really should avoid, since it is hacked up far beyond repair, but IT WORKS FINE as it is).

But we can do at least the clipping we need. It seems to work fine like this:

`int CharsetRendererV7::drawChar(byte *buffer, Common::Rect &clipRect, int x, int y, int pitch, int16 col, TextStyleFlags flags, byte chr) {`

	`if (!prepareDraw(chr))
		return 0;
	_width = getCharWidth(chr);
	if (_direction < 0)
		x -= _width;`

	int width = MIN(_origWidth, clipRect.right - x);
	int height = MIN(_origHeight, clipRect.bottom - y);

	_vm->_charsetColorMap[1] = col;
	VirtScreen &vs = _vm->_virtscr[kMainVirtScreen];
	drawBitsN(vs, buffer + (y + _offsY) * vs.pitch + x, _charPtr, *_fontPtr, y, width, height);
	return _direction * width;
}`

But it should be tested some more.

I have checked Hebrew which also seems fine (it seems to be just pure luck that the symbols under "Thrust" don't go over the Hebrew text, but they don't.

**Comment by AndywinXp** [comment]
Thanks so much athrxx, I'll do some testing and report back!
> (it seems to be just pure luck that the symbols under "Thrust" don't go over the Hebrew text, but they don't.

Could you please give me a pic example of that so that I can try to stress test that part?

**Comment by AndywinXp** [comment]
Hmm, while it does work fine for DIG, it seemingly breaks FT on that edge case I mentioned (notice the garbled exclamation point):
scummvm-ft-00001

**Comment by athrxx** [comment]
Sure...

Not that we could do much about it, anyway (except with some ugly hack). The problem with the Hebrew is that the coordinates weren't adapted to RTL when the fan translation was made. The x-coord still points to the left of the text instead of to the right. So all we can do is calculate the RTL x position by subtracting the x value from the screen width...

**Comment by athrxx** [comment]
For the broken FT screen shot: That just means that the clipping needs to be fixed. Maybe it is necessary to go inside drawBitsN(), although I'd rather avoid it. Could you supply a save file for that scene, please?

**Comment by AndywinXp** [comment]
Thanks for the screenshot!

[ft.zip](https://github.com/scummvm/scummvm/files/8357836/ft.zip)
Here's the savegame you asked. (FT Eng, Ver B)

**Comment by athrxx** [comment]
With this code it seems to work. I can have a '!' on the most right pixel of the screen.

I just got rid of the call to drawBitsN()...

`int CharsetRendererV7::drawChar(byte *buffer, Common::Rect &clipRect, int x, int y, int pitch, int16 col, TextStyleFlags flags, byte chr) {`
`
if (!prepareDraw(chr))
	return 0;
`
`_width = getCharWidth(chr);`

`if (_direction < 0)
	x -= _width;
`
	`int width = MIN(_origWidth, clipRect.right - x);
	int height = MIN(_origHeight, clipRect.bottom - y);`

	_vm->_charsetColorMap[1] = col;
	byte *cmap = _vm->_charsetColorMap;
	const byte *src = _charPtr;
	byte *dst = buffer + (y + _offsY) * pitch + x;
	uint8 bpp = *_fontPtr;
	byte bits = *src++;
	byte numbits = 8;
	pitch -= _origWidth;

	while (height--) {
		for (int dx = x; dx < x + _origWidth; ++dx) {
			byte color = (bits >> (8 - bpp)) & 0xFF;
			if (color && dx >= 0 && dx < x + width && y >= 0)
				*dst = cmap[color];
			dst++;
			bits <<= bpp;
			numbits -= bpp;
			if (numbits == 0) {
				bits = *src++;
				numbits = 8;
			}
		}
		dst += pitch;
		++y;
	}

	return _direction * width;
`}`

**Comment by AndywinXp** [comment]
Great job as always, this appears to work fine! Want me to do another full round of tests for FT (full and demo) and DIG demo just to be safe?

**Comment by athrxx** [comment]
A bit more testing is certainly a good idea...

About that thing with the text speed in FT (with "-b 400 ft"), what exactly is the issue? The fact that the text is faster than the speech?

Are you sure that this is different in the original?

**Comment by AndywinXp** [comment]
Precisely, the text is too quick. I would assume it is a textspeed ScummVM setting derived thing but then I don't know why it doesn't happen in 2.5.1 or in 2.6.0git (they both yield the same results as the interpreter)... Have I inadvertedly broken something when adding support for FT in this PR?

**Comment by athrxx** [comment]
Okay, one thing I see in disasm is that we don't seem to set
`_talkDelay = VAR(VAR_DEFAULT_TALK_DELAY);`

If you add that after line 643 (`if (usingOldSystem) {`),
does it solve the issue?

**Comment by AndywinXp** [comment]
...that's exactly it, it definitely had slipped under my radar. Thanks :-)

**Comment by AndywinXp** [comment]
> Sure...
> 
> Not that we could do much about it, anyway (except with some ugly hack). The problem with the Hebrew is that the coordinates weren't adapted to RTL when the fan translation was made. The x-coord still points to the left of the text instead of to the right. So all we can do is calculate the RTL x position by subtracting the x value from the screen width...

Is this fan translation supposed to look different in the interpreter than how we're displaying it right now? Or is it ScummVM-only, somehow?

**Comment by athrxx** [comment]
Fortunately, it is a completely ScummVM based project. It won't run with any original interpreter. So we can't break anything against the original ;-)

I can't read it either. But at least I can see whether it renders smoothly.

**Comment by AndywinXp** [comment]
> It won't run with any original interpreter. So we can't break anything against the original ;-)

Phew! That's good to know :-P

Very good, I'll do some more testing on FT and DIG demo during the next two days and hopefully everything works perfectly now.

**Comment by dwatteau** [comment]
Thanks to both of you, I confirm that the Asteroid Lander is now OK in my French version. The intro is still identical to the original interpreter one (the French version misplaces quite a lot of credit lines, which are now "properly" truncated or stacked, as in the original).

BTW, there's a small typo here:

```diff
diff --git a/engines/scumm/string_v7.h b/engines/scumm/string_v7.h
index 25fbd213c42..9ff57c1b078 100644
--- a/engines/scumm/string_v7.h
+++ b/engines/scumm/string_v7.h
@@ -20,7 +20,7 @@
  */
 
 #ifndef SCUMM_STRING_V7_H
-#define SCUMM_STRINGV_7_H
+#define SCUMM_STRING_V7_H
 
 #ifdef ENABLE_SCUMM_7_8
 
```

**Comment by AndywinXp** [comment]
> (the French version misplaces quite a lot of credit lines, which are now "properly" truncated or stacked, as in the original).

Oh the care they put into the localized versions of this game :-) You're not alone...
I still have to take a look at the DiMUSE workaround for SMUSH audio, I promise to do it this week; though don't hold your breath: that part of the system barely holds up in the original, as you could imagine, and for every workaround I may insert, it might break two other random video related things.

> BTW, there's a small typo here:
> 
> ```diff
> diff --git a/engines/scumm/string_v7.h b/engines/scumm/string_v7.h
> index 25fbd213c42..9ff57c1b078 100644
> --- a/engines/scumm/string_v7.h
> +++ b/engines/scumm/string_v7.h
> @@ -20,7 +20,7 @@
>   */
>  
>  #ifndef SCUMM_STRING_V7_H
> -#define SCUMM_STRINGV_7_H
> +#define SCUMM_STRING_V7_H
>  
>  #ifdef ENABLE_SCUMM_7_8
>  
> ```

Whoops! Thanks.

**Comment by BLooperZ** [comment]
If it helps, maybe I can update the hebrew translation to use the discovered RTL flag🤔
(or change other things if needed)

**Comment by athrxx** [comment]
> 

Well, like I said, we don't have to be faithful to any original here. So, as long as it looks fine, you don't have to make changes (I guess, the mirroring of the x coord works well enough in 99,x% of cases).

File: engines/scumm/string_v7.cpp [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+			x += _gr->drawChar(buffer, clipRect, x, y, pitch, col, flags, str[i]);
+		}
+		--numBytesMax;
+	}
+}
+
+#define SCUMM7_MAX_STRINGS		80
+
+void TextRenderer_v7::drawString(const char *str, byte *buffer, Common::Rect &clipRect, int x, int y, int pitch, int16 col, TextStyleFlags flags) {
+
```
Comment (athrxx) [inline engines/scumm/string_v7.cpp:182]: Please remove line 182 :-)

Comment (AndywinXp) [reply engines/scumm/string_v7.cpp]: Done!

**Comment by athrxx** [comment]
Quite surprisingly, the clipping also works for RTL/Hebrew. I was almost sure that this would require some more tweaking, but apparently it doesn't....

For testing purposes I spammed some '!' to the very left to see that we can have a '!' exactly at pixel 0 (with the spare/empty pixels to the left of the actual character being clipped).

**Comment by athrxx** [comment]
I am done playing through English COMI. No issues with the text...

**Comment by AndywinXp** [comment]
Thanks! I hope to complete my share of tests between today and tomorrow (and check italian versions too this time...)

**Comment by AndywinXp** [comment]
Demos are tested and good to go; a first glance at eng and localized versions of FT and DIG show that everything appears to be working as it should. I'll test the full games tomorrow.

**Comment by AndywinXp** [comment]
Just as a reminder, this (the "o" at the end being truncated) _is supposed to happen_, as the Aaron Giles interpreter shows on the right:
image

**Comment by athrxx** [comment]
Yes, this is something I also noticed.

**Comment by AndywinXp** [comment]
I rebased my branch on the current upstream master, it was getting quite difficult to test FT ita with DiMUSE crashing every 10 minutes :-)

**Comment by athrxx** [comment]
If there is no objection within the next 48h I would merge this.

**Comment by bluegr** [comment]
I tried playing around with these changes in FT and COMI and didn't notice anything odd. Overall, the changes look great! +1 from me :)

**Comment by athrxx** [comment]
Okay, the time has come to merge this... :-)

**Comment by AndywinXp** [comment]
Yahoo! :-D Great job!
